File: go/2023/02/day02.go
package main

import (
	_ "embed"
	"fmt"
	"strconv"
	"strings"
)

//go:embed input.txt
var input string

const (
	MAX_RED   = 12
	MAX_GREEN = 13
	MAX_BLUE  = 14
)

func part1(input string) int {
	lines := strings.Split(input, "\n")

	var sum int
	for lineIndex, line := range lines {
		_, line, _ = strings.Cut(line, ": ")

		isPossible := true
		var start int
		for i := 1; i < len(line); i++ {
			if line[i] == ' ' {
				amount, _ := strconv.Atoi(line[start:i])

				var r, g, b int
				switch line[i+1] {
				case 'r':
					r = amount
					i += 6
					start = i

				case 'g':
					g = amount
					i += 8
					start = i

				case 'b':
					b = amount
					i += 7
					start = i
				}

				if r > MAX_RED || g > MAX_GREEN || b > MAX_BLUE {
					isPossible = false
					break
				}
			}
		}

		if isPossible {
			sum += lineIndex + 1
		}
	}

	return sum
}

func part2(input string) int {
	lines := strings.Split(input, "\n")

	var sum int
	for _, line := range lines {
		_, line, _ = strings.Cut(line, ": ")

		var start, r, g, b int
		for i := 1; i < len(line); i++ {
			if line[i] == ' ' {
				amount, _ := strconv.Atoi(line[start:i])

				switch line[i+1] {
				case 'r':
					r = max(r, amount)
					i += 6
					start = i

				case 'g':
					g = max(g, amount)
					i += 8
					start = i

				case 'b':
					b = max(b, amount)
					i += 7
					start = i
				}
			}
		}

		sum += r * g * b
	}

	return sum
}

func main() {
	fmt.Println("--- 2023 day 02 answer ---")
	fmt.Println("part 1:\t", part1(input))
	fmt.Println("part 2:\t", part2(input))
}

File: c/2023/02/day02.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int part1(FILE *f) {
  const int MAX_R = 12, MAX_G = 13, MAX_B = 14;
  int gameIndex = 1, sum = 0;

  char line[256];
  while (fgets(line, sizeof(line), f)) {
    bool isValid = true;

    char *i = strchr(line, ':');
    while (*i != '\n' && *i != '\0') {
      i += 2;
      int value = atoi(i);
      i = strchr(i, ' ') + 1;
      switch (*i) {
        case 'r':
          if (value > MAX_R) isValid = false;
          i += 3;
          break;
        case 'g':
          if (value > MAX_G) isValid = false;
          i += 5;
          break;
        case 'b':
          if (value > MAX_B) isValid = false;
          i += 4;
          break;
      }
      if (!isValid) break;
    }

    if (isValid) sum += gameIndex;
    gameIndex++;
  }

  return sum;
}

int part2(FILE *f) {
  int sum = 0;

  char line[256];
  while (fgets(line, sizeof(line), f)) {
    int max_r = 0, max_g = 0, max_b = 0;

    char *i = strchr(line, ':');
    while (*i != '\n' && *i != '\0') {
      i += 2;
      int value = atoi(i);
      i = strchr(i, ' ') + 1;
      switch (*i) {
        case 'r':
          if (value > max_r) max_r = value;
          i += 3;
          break;
        case 'g':
          if (value > max_g) max_g = value;
          i += 5;
          break;
        case 'b':
          if (value > max_b) max_b = value;
          i += 4;
          break;
      }
    }

    sum += max_r * max_g * max_b;
  }

  return sum;
}

int main(void) {
  FILE *f = fopen("input.txt", "r");
  if (f == NULL) {
    perror("fopen error");
    exit(1);
  }

  printf("--- 2023 day 01 answer ---\n");
  printf("part 1:\t%d\n", part1(f));
  rewind(f);
  printf("part 2:\t%d\n", part2(f));

  fclose(f);
  return 0;
}
